mdn/src/main/java/org/apache/james/mdn/fields/Gateway.java

```diff
@@ -29,21 +30,44 @@
 public class Gateway implements Field {
     public static final String FIELD_NAME = "MDN-Gateway";
 
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder {
+        private Optional<AddressType> nameType;
+        private Text name;
+
+        private Builder() {
+            this.nameType = Optional.empty();
+        }
+
+        public Builder nameType(AddressType nameType) {
+            this.nameType = Optional.of(nameType);
+            return this;
+        }
+
+        public Builder name(Text name) {
+            this.name = name;
+            return this;
+        }
+
+        public Gateway build() {
+            return new Gateway(nameType.orElse(AddressType.DNS), name);
+        }
+    }
+
     private final AddressType nameType;
     private final Text name;
 
-    public Gateway(AddressType nameType, Text name) {
+    private Gateway(AddressType nameType, Text name) {
         Preconditions.checkNotNull(nameType);
```

Comment:
You can move preconditions in the `Builder.build()` method.
(same comment for other builders)

Reply:
Why? I think having early errors is valuable...

Reply:
done. @chibenwa I don't understand your comment, it changes nothing to errors handling, no?

Reply:
Sorry I misread.

Reply:
It's a pattern we are using.
Plus, you don't have to create an instance if you know earlier it will fail.